cli/find_in_files.php

```diff
@@ -0,0 +1,121 @@
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * Search strings throughout all texts in the whole database.
+ *
+ * @package    tool_advancedreplace
+ * @copyright  2024 Catalyst IT Australia Pty Ltd
+ * @license    http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+use tool_advancedreplace\file_search;
+
+define('CLI_SCRIPT', true);
+
+require(__DIR__.'/../../../../config.php');
+require_once($CFG->libdir.'/clilib.php');
+require_once($CFG->libdir.'/adminlib.php');
+
+$help =
+    "Search for text in moodle files.
+
+Options:
+--search=STRING                  Required if --regex-match is not specified. String to search for.
+--regex-match=STRING             Required if --search is not specified. Use regular expression to match the search string.
+--output=FILE                    Required output file. If not specified, output to stdout.
+--components=componentname:areaname    Components and areas to search. Separate multiple components/areas with a comma.
+                                 If not specified, search all tables and columns.
+                                 If specify table only, search all columns in the table.
+                                 Example:
+                                    --components=core_h5p:content
+                                    --components=core_h5p,assign_submission:submission
+--skip-components=componentname   Components to skip. Separate multiple components with a comma.
+                                 Example:
+                                    --skip-components=core_h5p,assign_submission
+--mimetypes=mimetype,mimetype   Mimetypes to be searched. Separate multipler type with commas.
+                                If empty, all mimetypes will be considred.
+--skip-mimetypes=mimetype,mimetype Mimetypes to be skipped.
+--filenames=filename            Cooma-separated list of file names to be searched.
+--skip-filenames=filename       Comma-separated list of file names to be omitted from the search.
+--skip-areas=areaname        Areas to skip. Separate multiple areas with a comma.
+                                 Example:
+                                    --skip-areas=draft,export
+-h, --help                       Print out this help.
+
+Example:
+\$ php find_in_files.php --regex-match=thelostsoul\\d+ --output=/tmp/result.csv
+\$ php find_in_files.php --regex-match=doris\\d+ --output=/tmp/result.csv --mimetype=application/xip.h5p
+";
+
+list($options, $unrecognized) = cli_get_params(
+    [
+        'regex-match'   => null,
+        'output'        => null,
+        'components'    => '',
+        'skip-components'   => '',
+        'mimetypes'     => '',
+        'skip-mimetypes' => '',
+        'filenames'     => '',
+        'skip-filenames' => '',
+        'skip-areas'    => '',
+        'help'          => false,
+    ],
+    [
+        'h' => 'help',
+    ]
+);
+core_php_time_limit::raise();
```

Comment:
I think this also needs `raise_memory_limit(MEMORY_HUGE);
`
I was running into issues when testing with CLI because the fileset was too large.

Even with that the search was crashing locally after processing a touch over 520,000 records - granted the filters will drastically lower that, but some other techniques may need to be used to help with this.